### PR TITLE
Logical operations

### PR DESCRIPTION
**Pull request by nishnik**
This PR has implementations of Logical `And`, `Or` and `Not`.

### CODE REVIEW

File: symengine/logic.cpp [hunk 1]
```diff
@@ -146,4 +146,103 @@ int Piecewise::compare(const Basic &o) const
...
+
+int Or::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Or>(o))
+    return unified_compare(args, static_cast<const Or &>(o).args);
+}
+
+bool Or::is_canonical(const set_boolean &container)
+{
+    return container.size() >= 2;
```
Comment (isuruf) [inline symengine/logic.cpp]: You need to check that none of the elements in container are of type `Or`

File: symengine/logic.h [hunk 2]
```diff
@@ -96,6 +96,128 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
...
+    }
+    set_boolean flattened_args;
+    for (auto &a : args) {
+        if (is_a<caller>(*a)) {
+            const caller &to_insert = static_cast<const caller &>(*a);
+            flattened_args.insert(to_insert.args.begin(), to_insert.args.end());
+        } else {
+            flattened_args.insert(a);
+        }
+    }
```
Comment (isuruf) [inline symengine/logic.h]: You don't need two sets, `args` and `flattened_args` right? You can check whether `a` is a `caller` in Line 163 and insert the elements to `args`

File: symengine/logic.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/logic.cpp]: You also need to check that the none of the elements are `BooleanAtom`s. Basically, `is_canonical` should do all the checks that you do when you are creating the object. In this case, `is_canonical` should check all the cases that `and_or` does

File: symengine/logic.h [hunk 3]
```diff
@@ -96,6 +96,128 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
...
+            s.insert(logical_not(a));
+        }
+        return logical_and(s);
+    } else if (is_a<And>(*s)) {
+        const And &o = static_cast<const And &>(*s);
+        set_boolean s;
+        for (auto &a : o.args) {
+            s.insert(logical_not(a));
+        }
+        return logical_or(s);
```
Comment (isuruf) [inline symengine/logic.h]: This doesn't need all the checking right? You can create a `Or` object directly here.

File: symengine/logic.h [hunk 4]
```diff
@@ -96,6 +96,128 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
...
+{
+    return and_or<And>(s, false);
+}
+
+inline RCP<const Boolean> logical_or(const set_boolean &s)
+{
+    return and_or<Or>(s, true);
+}
+
+inline RCP<const Boolean> logical_not(const RCP<const Boolean> &s)
```
Comment (isuruf) [inline symengine/logic.h]: Can you move this to the implementation file?

File: symengine/logic.h [hunk 5]
```diff
@@ -96,6 +96,128 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
...
+    virtual vec_basic get_args() const;
+    virtual bool __eq__(const Basic &o) const;
+    //! Structural equality comparator
+    virtual int compare(const Basic &o) const;
+};
+
+RCP<const Boolean> logical_not(const RCP<const Boolean> &s);
+
+template <typename caller>
+RCP<const Boolean> and_or(const set_boolean &s, const bool &op_x_notx)
```
Comment (isuruf) [inline symengine/logic.h]: Can you move the definition and the implementation to the `.cpp` file?

File: symengine/logic.h (hunk 3 above)
Comment (nishnik) [inline symengine/logic.h]: `s` might be a `Contains` object.

File: symengine/logic.cpp [hunk 6]
```diff
@@ -146,4 +146,186 @@ int Piecewise::compare(const Basic &o) const
...
+            s.insert(logical_not(a));
+        }
+        return logical_and(s);
+    } else if (is_a<And>(*s)) {
+        const And &o = static_cast<const And &>(*s);
+        set_boolean s;
+        for (auto &a : o.args) {
+            s.insert(logical_not(a));
+        }
+        return logical_or(s);
```
Comment (isuruf) [inline symengine/logic.cpp]: What do you mean by, `s might be a Contains object.`? `s` is a `set_boolean` right?

File: symengine/logic.h [hunk 7]
```diff
@@ -96,6 +96,71 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
...
+    //! \return the hash
+    std::size_t __hash__() const;
+    virtual vec_basic get_args() const;
+    virtual bool __eq__(const Basic &o) const;
+    //! Structural equality comparator
+    virtual int compare(const Basic &o) const;
+};
+
+template <typename caller>
+RCP<const Boolean> and_or(const set_boolean &s, const bool &op_x_notx);
```
Comment (isuruf) [inline symengine/logic.h]: You have to move this to the implementation file as it is not a good idea to keep only the definition of a templated functions in a header file.

File: symengine/logic.h [hunk 8]
```diff
@@ -96,6 +96,71 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
     return make_rcp<Piecewise>(std::move(vec));
 }
 
+class And : public Boolean
+{
+public:
```
Comment (isuruf) [inline symengine/logic.h]: Can you make this private?

File: symengine/logic.h [hunk 9]
```diff
@@ -96,6 +96,71 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
...
+    virtual vec_basic get_args() const;
+    virtual bool __eq__(const Basic &o) const;
+    //! Structural equality comparator
+    virtual int compare(const Basic &o) const;
+};
+
+class Not : public Boolean
+{
+public:
+    RCP<const Boolean> arg;
```
Comment (isuruf) [inline symengine/logic.h]: Same here

File: symengine/logic.h [hunk 10]
```diff
@@ -96,6 +96,71 @@ inline RCP<const Basic> piecewise(PiecewiseVec &&vec)
...
+    std::size_t __hash__() const;
+    virtual vec_basic get_args() const;
+    virtual bool __eq__(const Basic &o) const;
+    //! Structural equality comparator
+    virtual int compare(const Basic &o) const;
+};
+
+class Or : public Boolean
+{
+public:
```
Comment (isuruf) [inline symengine/logic.h]: Same here

File: symengine/tests/logic/test_logic.cpp [hunk 11]
```diff
@@ -86,3 +90,97 @@ TEST_CASE("Piecewise", "[logic]")
...
+{
+    auto x = symbol("x");
+    auto int1 = interval(integer(1), integer(2), false, false);
+    auto int2 = interval(integer(1), integer(5), false, false);
+    auto c1 = contains(x, int1);
+    auto c2 = contains(x, int2);
+
+    REQUIRE(eq(*logical_not(boolTrue), *boolFalse));
+    REQUIRE(eq(*logical_not(boolFalse), *boolTrue));
+    REQUIRE(logical_not(c1)->__str__() == "Not(Contains(x, [1, 2]))");
```
Comment (isuruf) [inline symengine/tests/logic/test_logic.cpp]: For this, maybe `Contains(x, (-oo, 1) U (2, oo))`  is better ?

Comment (isuruf) [inline symengine/tests/logic/test_logic.cpp]: On second thought, this is fine.

**Pull request merged**